tools/cli/helpers/promptForProject.js

```diff
@@ -45,3 +42,146 @@ export async function promptForProject( options ) {
+ */
+export async function promptForGenerate( options ) {
+	let typeAnswer;
+	let nameAnswer;
+	let tries = 0;
+
+	// Get project type if not passed as an option.
+	if ( ! options.type || options.type.length === 0 ) {
+		typeAnswer = await promptForType();
+	} else if ( ! projectTypes.includes( pluralize( options.type ) ) ) {
+		return new Error( 'Must be a valid project type' );
+	}
+
+	// Get the appropriate list of project prompts based on type
+	const questions = getQuestions( options.type || typeAnswer.type );
+	if ( ! questions ) {
+		return new Error( "Sorry! That's not supported yet!" );
+	}
+
+	// Validate name if it was passed as an option.
+	if ( options.name ) {
+		checkNameValid( options.type || typeAnswer.type, options.name ) ? nameAnswer = options.name : nameAnswer = null;
+	}
+
+	// Keep asking for name if it fails validation
+	while ( ! nameAnswer ) {
+		try {
+			tries++;
+			const rawNameAnswer = await promptForName();
+			checkNameValid( options.type || typeAnswer.type, rawNameAnswer.name ) ? nameAnswer = rawNameAnswer : nameAnswer = null;
+		} catch ( err ) {
+			if ( tries >= 3 ) {
+				console.error( 'You are really struggling here. Might be time to take a walk.' );
+				console.error( err.name + ': ' + err.message );
+			}
+		}
+	}
+
+	// Give the list of questions
+	const finalAnswers = await inquirer.prompt( questions );
+
+	return {
+		...options,
+		type: pluralize.singular( options.type || typeAnswer.type ),
+		name: options.name || nameAnswer.name,
+		...finalAnswers,
+	};
+}
+
+/**
+ * Prompt for type.
+ *
+ * If no type is passed via `options`, then it will prompt for the type of project.
+ *
+ * @param {object} options - Passthrough of an object, meant to accept argv.
+ *
+ * @returns {object} object with the type property appended.
+ */
+export async function promptForType( options = { type: '' } ) {
+	let typeAnswer;
+
+	if ( ! options.type || options.type.length === 0 ) {
+		typeAnswer = await inquirer.prompt( {
+			type: 'list',
+			name: 'type',
+			message: 'What type of project are you working on today?',
+			choices: projectTypes,
+		} );
+	} else if ( ! projectTypes.includes( options.type ) ) {
+		return new Error( 'Must be an accepted project type.' );
+	}
+
+	return {
+		...options,
+		type: options.type || typeAnswer.type,
+	};
+}
+
+/**
+ * Prompt for new project name.
+ *
+ * If no name is passed via `options`, then it will prompt for the name of project.
+ *
+ * @param {object} options - Passthrough of an object, meant to accept argv.
+ *
+ * @returns {object} object with the name property appended.
+ */
+export async function promptForName( options = { name: '' } ) {
+	let nameAnswer;
+
+	if ( ! options.name || options.name.length === 0 ) {
+		nameAnswer = await inquirer.prompt( {
+			type: 'input',
+			name: 'name',
+			message: 'What is your project called?',
+		} );
+	}
+
+	return {
+		...options,
+		name: options.name || nameAnswer.name,
```

Comment:
Is there a particular reason we need to return an object here instead of just the nameAnswer.name value (same question for promoptForType above)? We create the object in promptForGenerate and promptForProject, so thinking we can just return the single value and then can clean up the assignments to not have to take both object or string values into account.

Reply:
the original idea was to mirror promptforproject and the usage seen here: https://github.com/Automattic/jetpack/blob/master/tools/cli/commands/install.js#L61

Consuming code can just pass it the argv and it'll return the same thing back, just with a name after prompting if wasn't already declared.

It makes it cleaner when we want to support `--name` flags IMO.